### ISSUE TITLE
False positive for SC2154

### ISSUE THREAD

**Issue by mcandre** [koalaman/shellcheck#1046]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154
- My shellcheck version (`shellcheck --version` or "online"): 0.4.6
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
# For your shebangs
#
# mkdir -p ~/bin
# cp .../shmatlab% ~/bin/shmatlab%
#

code=<<END
varargin = regexp('${1+"$@"}', ' ', 'split');
nvarargin = length(varargin);
run('$1');
exit
END

matlab \
    -nojvm \
    -nodisplay \
    -nosplash \
    -r "$code" |
    tail -n +16
```

#### Here's what shellcheck currently says:

```console
$ shellcheck shmatlab%
In shmatlab% line 19:
    -r "$code" |
        ^-- SC2154: code is referenced but not assigned.
```

#### Here's what I wanted or expected to see:

(SC2154 should not appear. The heredoc will assign a value to the `code` variable.)

**Comment by kurahaupo** [comment]
On Mon, 13 Nov 2017, Andrew Pennebaker wrote:
> (SC2154 should not appear. The heredoc will assign a value to the `code` variable.)

I would recommend testing one's code to verify that it actually does what
one asserts that it does, before filing a bug report claiming that
shellcheck has made a mistake.

The << token is a redirection operator, not an expansion operator.

In point of fact, shellcheck *has* made a mistake; "code" HAS been assigned
a value, just not what you think has been assigned.

**Comment by mcandre** [comment]
`>>` operates as redirection, `<<` operates as heredoc:

https://www.esrl.noaa.gov/gmd/dv/hats/cats/stations/qnxman/sh.html

**Comment by jleffler** [comment]
I think you've misunderstood what's going on.  The `<<` operator is an input redirection operator; it provides standard input to a command.

Consider this script — and adaptation of your example:

    #!/bin/bash

    echo "Part 1"
    variable=<<EOF
    line1
    line2
    EOF

    echo "variable=[$variable]"

    echo "Part 2"
    variable=<<EOF sed 's/^/--: /'
    line1
    line2
    EOF

    echo "variable=[$variable]"

    echo "Part 3"
    <<EOF bash -c 'set -u; echo variable=$variable' xxx
    line1
    line2
    EOF

    variable=<<EOF bash -c 'set -u; echo variable=$variable' xxx
    line1
    line2
    EOF

    echo "variable=[$variable]"

When I run it on a Mac running macOS High Sierra 10.13.1 using Bash 3.25 (as provided by Apple), I get:

    Part 1
    variable=[]
    Part 2
    --: line1
    --: line2
    variable=[]
    Part 3
    xxx: variable: unbound variable
    variable=
    variable=[]

Part 1 shows that nothing is set in the variable.  Part 2 shows that the here document is available as standard input to the command.  Part 3 shows that `variable` is set as an empty environment variable to the command that's executed.  Nothing would change in any of the examples if you used:

    variable=  <<EOF

with space between the `=` and the redirection `<<`.

Yes, I/O redirection can precede the command.  It's aconventional, but permissible.

(I had a 'stellar' example at work:

    <something> another sed 's/<abc>/<def>/'

The `<something` redirects standard input; the `> another` redirects standard output; the command follows both redirections.  No; it wasn't what I wrote; I just spotted it and had a giggle.)
